pkg/machinery/config/merge/merge_test.go

```diff
@@ -0,0 +1,254 @@
+
+	return nil
+}
+
+func TestMerge(t *testing.T) {
+	for _, tt := range []struct {
+		name        string
+		left, right Config
+		expected    Config
+	}{
+		{
+			name: "zero",
+		},
+		{
+			name: "partial merge",
+			left: Config{
+				A: "a",
+				B: 3,
+				C: pointer.To(true),
+				Slice: []Struct{
+					{
+						DA: true,
+						DB: pointer.To(1),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DA: true,
+					},
+					"b": {
+						DB: pointer.To(2),
+					},
+				},
+			},
+			right: Config{
+				A: "aa",
+				B: 4,
+				Slice: []Struct{
+					{
+						DA: false,
+						DB: pointer.To(2),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DB: pointer.To(3),
+					},
+					"b": {
+						DA: true,
+						DB: pointer.To(5),
+					},
+					"c": {
+						DB: pointer.To(4),
+					},
+				},
+			},
+			expected: Config{
+				A: "aa",
+				B: 4,
+				C: pointer.To(true),
+				Slice: []Struct{
+					{
+						DA: true,
+						DB: pointer.To(1),
+					},
+					{
+						DA: false,
+						DB: pointer.To(2),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DB: pointer.To(3),
+					},
+					"b": {
+						DA: true,
+						DB: pointer.To(5),
+					},
+					"c": {
+						DB: pointer.To(4),
+					},
+				},
+			},
+		},
+		{
+			name: "merge with zero",
+			left: Config{
+				A: "a",
+				B: 3,
+				C: pointer.To(true),
+				Slice: []Struct{
+					{
+						DA: false,
+						DB: pointer.To(2),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DA: true,
+					},
+					"b": {
+						DB: pointer.To(2),
+					},
+				},
+			},
+			right: Config{},
+			expected: Config{
+				A: "a",
+				B: 3,
+				C: pointer.To(true),
+				Slice: []Struct{
+					{
+						DA: false,
+						DB: pointer.To(2),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DA: true,
+					},
+					"b": {
+						DB: pointer.To(2),
+					},
+				},
+			},
+		},
+		{
+			name: "merge from zero",
+			left: Config{},
+			right: Config{
+				A: "a",
+				B: 3,
+				C: pointer.To(true),
+				Slice: []Struct{
+					{
+						DA: false,
+						DB: pointer.To(2),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DA: true,
+					},
+					"b": {
+						DB: pointer.To(2),
+					},
+				},
+			},
+			expected: Config{
+				A: "a",
+				B: 3,
+				C: pointer.To(true),
+				Slice: []Struct{
+					{
+						DA: false,
+						DB: pointer.To(2),
+					},
+				},
+				Map: map[string]Struct{
+					"a": {
+						DA: true,
+					},
+					"b": {
+						DB: pointer.To(2),
+					},
+				},
+			},
+		},
+		{
+			name: "replace slice",
+			left: Config{
+				ReplacedSlice: []string{"a", "b"},
+			},
+			right: Config{
+				ReplacedSlice: []string{"c", "d"},
+			},
+			expected: Config{
+				ReplacedSlice: []string{"c", "d"},
+			},
+		},
+		{
+			name: "zero slice",
+			left: Config{},
+			right: Config{
+				Slice: []Struct{},
+			},
+			expected: Config{
+				Slice: []Struct{},
+			},
+		},
+		{
+			name: "custom slice",
```

Comment:
Should we also have a case when `left` is not settable?

Reply:
I don't think it's possible actually with the structs we have... it's like merging into an `int`, for structs/maps/slices, we always have a way to set the left-hand part. I added the checks mostly to avoid panics, but it should never happen under normal operations